### ISSUE TITLE
Stockfish doesn't recognize relay checks

### ISSUE THREAD

**Issue by mnemi** [mnemi/RelayChess#9]
It seems that Stockfish will suggest an illegal move and error out when the move it tries to play would put one of the kings in check by a relayed piece.

Example steps to reproduce:
- Play the AI on level 1
- 1. e4 d5 2. exd5 Qxd5 3. Nf3 Qd6 4. Bc4 Nf6 5. Na3 Nc6 6. B(N)xd6
Stockfish will then try to play the move 6... exd6, but that would put its king in check from White's king who has the relay powers of a rook. I guess Stockfish doesn't use a piece's relay attributes when calculating whether or not a move would put it in check.

Maybe @ddugovic could help out here?

**Comment by ddugovic** [comment]
I beautified `stockfish.js` so we may start to fix the bug:
https://gist.github.com/ddugovic/fcb01fe12e189ec4a30f86b467e5658f

**Comment by mnemi** [comment]
Unfortunately that is just prettified, but the file is still minified (variable names are mangled, etc.).

I haven't really looked into the stockfish side of this codebase much, but I should've mentioned this is just using regular stockfish, i.e. https://github.com/niklasf/stockfish.js . Unfortunately, it will be hard to fix these bugs because you actually __removed__ the Relay variant earlier this year hahaha! 
See: https://github.com/niklasf/stockfish.js/commit/ac56608f6ecfbdc447a06b9de6fbf512b71ba945
& https://github.com/ddugovic/Stockfish/commit/ac56608f6ecfbdc447a06b9de6fbf512b71ba945

**Comment by ddugovic** [comment]
Yeah, I removed the Relay variant because other than the `is_relay()` function it was 0% developed.  So I'm very confused how `stockfish.js` plays relay chess.

**Comment by mnemi** [comment]
Wait, really? Then how the heck did the original author get Stockfish working? Ok, this is weird.
Two theories come to mind:
1. He modified Stockfish himself and compiled it by hand with emscripten (unlikely, and a lot of work, but possible)
2. Is it possible to hook into Stockfish and provide it a custom movegen? If so, maybe he gave Stockfish the modified move gen from chess.js. That would explain why at all levels the AI is blind the relay rules when it's not their move. This makes sense because it won't notice a move puts it in check or that the queen is attacked by a relayed bishop with knight powers. So maybe Stockfish is just playing regular chess, but using relay movegen so when it tries to move a piece it will use relay moves given to it if they seem best for normal chess?

**Comment by ddugovic** [comment]
Huh... looking at the code, despite many hours of looking I forget how this was coded, but apparently it's related:
https://github.com/ddugovic/Stockfish/commit/ac56608f6ecfbdc447a06b9de6fbf512b71ba945#diff-85fbbdc0ae2caf8c12bb6a676da65c9aR379

At any rate it's probably best that I fork Rustfish since hopefully Rust somehow makes it easier to correctly implement.

**Comment by mnemi** [comment]
I can try and take a stab at Stockfish, but chess engines and c++ aren't even close to my forte. If you could point me in the right direction that'd be much appreciated :smile:. How is that implementation calculating piece attributes? Not sure if I'm phrasing that properly, but in the chess.js version each square of the 0x88 board array has either `null` or a `piece` object and each piece has an attributes field which is just an `int`. When calculating the pieces attributes, it calculates the "attackers" of the same color and then `or`s them onto the attributes field.

**Comment by ddugovic** [comment]
Apparently I coded a thing which compiles:
https://github.com/ddugovic/Rustfish/commit/a3dd93873c16dbc1f747edce440df1e86180d76f

**Comment by mnemi** [comment]
An interesting finding: all levels of stockfish will leave their queen hanging in the given position and will generate an illegal move after.

This further suggests that Stockfish is actually playing regular chess, but using a "relay movegen". Might be why it was so easy to get Stockfish playing RelayChess.

**Comment by ddugovic** [comment]
Sadly after further effort, I discovered that attempting to support Relay Chess in Rustfish is infeasible due to functional design of the engine and semantics of the language.  It could be done but would be a monumental undertaking; but on the bright side, doing it in Rust would protect against subtle coding errors!

**Comment by ddugovic** [comment]
I wonder whether a Haskell chess engine might be easier to enhance...
https://github.com/nionita/Barbarossa

**Comment by mnemi** [comment]
I don't think either would be easier than at least trying to fix the Stockfish movegen. 

After doing some searching, I think the easiest way to fix movegen, but keep it compatible with Stockfish's bitboards, would be to have 1 array holding 64 n-bit unsigned numbers. This array would represent the squares on the board and each square's number would be the attributes/"relay powers" for that square. So if we decided that 0010 would represent a knight, then having that bit set would mean the current square has knight powers. This would make it fairly easy to reuse the existing `chess.js` logic without having to make a bitboard for each combination of piece powers (i.e. a knight-bishop board).

Also, we might be able to reuse the existing movegen for Stockfish. Just augment it with these attribute checks and reuse the existing routines for the given attribute (i.e. if a knight attribute is found for the square a bishop is on, then also generate moves for a knight from that square).

I think this is doable and worth a try before moving onto engines that aren't as well established and written in languages that would probably make this harder (especially Haskell. Working around immutable bitboards sounds like a PITA).

**Comment by ddugovic** [comment]
I'll see what I can do... https://github.com/ddugovic/Stockfish/tree/relay

I think the key there is: what updates the bitboard(s) and when?  At one point I'd considered that, but many functions make assumptions about checks and pinned pieces:
https://github.com/ddugovic/Stockfish/commit/ac56608f6ecfbdc447a06b9de6fbf512b71ba945#diff-85fbbdc0ae2caf8c12bb6a676da65c9aR379

For example consider position [8/4k3/8/8/1R5R/8/6PP/q3Q2K b - -](https://lichess.org/editor/8/4k3/8/8/1R5R/8/6PP/q3Q2K_b_-_-).  Black is in triple-check, and normally if a king is in check it isn't possible to capture multiple checking pieces.  Also consider that preventing the relay of a move (or worse, exposing the relay of a move) potentially happens every time a piece moves out of the line of sight of a queen, rook, or bishop.

**Comment by mnemi** [comment]
Thanks for the example. That really put in perspective how difficult this might be. I still think it's doable, but I'll need to familiarize myself with Stockfish a bit before I make a pull request.

In the example, I think we're lucky here because Relay pieces don't relay moves so there's not some sort of exponential blowup in terms of what we need to check! When in check, could we see if the square attacking us has relay powers and if true, are the relay powers the ones doing the checking? If so, then we could regard that as a normal check, but from the base piece (the piece relaying it's powers to the relay piece). So capturing the base piece would count as a legal move to get out of check.

**Comment by mnemi** [comment]
Also, I think the example there would require the same additions as this one: https://lichess.org/editor/8/4k3/8/8/Q3B3/8/6PP/q5NK_b_-_-

It's really an example of how to deal with relay checks. Just in that special case the piece relaying the powers also happens to be giving check itself.

**Comment by ddugovic** [comment]
That's true, subject to an assumption that capturing the base piece is legal!
[8/4k1N1/8/8/1Q2B3/8/1q4PP/Q5NK b - -](https://lichess.org/editor/8/4k1N1/8/8/1Q2B3/8/1q4PP/Q5NK_b_-_-)

**Comment by ddugovic** [comment]
I created some helper functions although there's more work to be done:
https://github.com/ddugovic/Stockfish/commit/3e7de1246a9e82ae052c231fabbc978627ab6cf0

**Comment by mnemi** [comment]
I've been looking over stockfish and I'm very confused... but that's a personal problem if I've ever heard one :laughing:. I'll keep bashing my head against it.

In the meantime, what would be the bare minimum list of features necessary to get Relaychess in Stockfish?

- pseudo-legal movegen
  - relay attackers
  - sliding pieces reveal relay powers
  - Other complicated scenarios?
- legal movegen
- special eval for relayed pieces?

Just this list sounds really complicated to implement. I'm starting to lose hope in this as well :sad:. The `chess.js` implementation of movegen looks pretty well done, but it's using a 0x88 representation, not bitboards. I guess bitboards are generally harder to work with, regardless of variant?

**Comment by ddugovic** [comment]
Sorry, I don't mean to suggest that you'd be at home in Stockfish... the relay (sub)variant is "functional" in nature and Stockfish simply isn't.  Even if SF used 0x88 or some other representation I still think... well, it's like trying to convert an airplane into a boat.  It can be done but there might be other vehicles which are more boat-like or which might be simpler to work with.

The good news is that if you compile with `debug=yes` SF has numerous built-in assertions so simple tests like `./stockfish bench relay` will noisily fail (and `gdb ./stockfish` can provide backtraces).  Lately the strategy I've been using for SF variant development is "implement basic rules, then fix bugs as they are observed" and finally tune for performance.  Generally users have many ideas about special evaluations, etc. but 95% of those ideas don't result in a rating gain.

The `chess.js` implementation indeed looks pretty well done and seemed quite straightforward, which is why I've been looking everywhere other than SF to find an easy solution.
